### DISCUSSION TITLE
Interfaces and final classes for services and their dependencies

### DISCUSSION THREAD

**Discussion by JoshuaBehrens** [shopware/platform#3163] [RFC / Proposal]
**Effort**: middle

Interfaces can also be contract classes in this context!

### Description: 

Every time you publish a class you have like two APIs. The API how this class is instantiated and used as well as how you extend this class and re-use its inner components.

### Benefits

Clear distinction on code re-usage, improve and extend entrypoints for extending classes and changing behaviour within the framework.

### Break strategy

Have a clear list of classes that one might have extended already that you now should change to an interface extension.
Have a guide that explains that you have abstract classes that are intended only for re-usage, interfaces for defining services, and final classes that only service certain implementations.

**Comment by OliverSkroblin** [comment]
Hey @JoshuaBehrens, great that you write a proposal.

@shyim and i discussed your proposal and are not completly sure what exactly you mean 😉.

Could you give us a few examples what you exactly expect? (code would be great)

**Comment by JoshuaBehrens** [comment]
I would like to have an interface around the elasticsearch helper like the discussion we had in here https://github.com/shopware/platform/pull/2035 . I can provide further examples later on.

**Comment by JoshuaBehrens** [comment]
ok so e.g. you have three types of classes:

##### 1. interface/contract

You define structure for decoration or reimplementation. No "default" implementation. Just method signatures. You already have those: AbstractRoute classes are a good example for this.

##### 2. base classes

Classes that already ship an implementation where you only have to switch some of default ones. Like EntityDefinition. EntityDefinition would need an interface though because you can decorate it that way.

##### 3. implementation classes

Everything that implements an interface that is meant to describe a "service" and not a struct like the DAL entity classes would be final because they are "just" an implementation of an interface. They don't want to be extended. When you want someone to extend you you tell them by marking yourself as abstract. This way you don't have to worry about bad decorators anymore when you have a plan how to clearly mark decorators.

BUT! This is not a need for final classes. This is a need for final classes that extend an interface (or contract).

##### 4. Classes that would change

E.g. the RepositorySearchDetector. It has a static method (therefore-ish) no interface and you are unable to change that. I see that this is quite internally. So there is also HookableEventCollector. This is marked as internal but I need to decorate it to add custom entities to the app system in a "close to the standard"-solution. I can't do that properly. Without this change I don't see a chance for self-hosted-apps to integrate there properly without a decorator (and a change with that constant, but this is a different story).

**Comment by JanPietrzyk** [comment]
Hi @JoshuaBehrens 

Wouldn't this in turn make all classes publicly replaceable? We agreed against interfaces on every service and class for exactly that reason. 

Our problem is that backward compatible development in the sources is a must and therefore closing all inter class communication with an interface does not lead us anywhere. As you know, we try to provide events and in some places domain motivated contracts for that use case.

While I absolutely know that there are some use cases that necessitate deep changes in the underlying system - at least in my experience - the `final+interface` solution does not help as much as one would hope.

Or am I getting this wrong? :)

**Comment by rattananen** [comment]
I agree with @JoshuaBehrens.
This proposal is common implement OOP that make class easier to composed, implemented and understandable.
1. interface: it's signature of class. we can test what is this class or what this class able to do by interface (instanceof).
2. base classes (abstract class): It just example for how to implement.
3. implementation classes (concrete class): may extend from example or implement their own way by implement interface or we can create crazy omni class by implement multiple interface.

case:
1. Shopware\Core\Framework\DataAbstractionLayer\Entity: now when we want to create entity must extends this class but this class has many unnecessary property and method (some entity don't want timestamp or translation) this may cause trouble if we add or modify this class (see later).

2. Shopware\Core\Framework\Struct\ArrayEntity and Shopware\Core\Framework\DataAbstractionLayer\TranslationEntity: we can not compose class from both.

my example idea:
Shopware\Core\Framework\DataAbstractionLayer\Entity, Shopware\Core\Framework\Struct\ArrayEntity and Shopware\Core\Framework\DataAbstractionLayer\TranslationEntity change to
- EntityInterface (may extends JsonSerializable if mandatory): method getUniqueIdentifier?, getApiAlias, getInternalEntityName
- TranslatableInterface: methods need for translate.
- TranslatebleTrait: Here PHP provide cool tool for this purpose.
- VersionableInterface, VersionableTrait
- TimestampTrait, ArrayAccessTrait, PropertyAccessTrait

and then create example class: 
AbstractEntity implements EntityInterface: create example method from interface compose TimestampTrait.

now create MyEntity extends AbstractEntity
and another NoNeedTimestampEntity implements EntityInterface
and another NeedTranlateEntity extends AbstractEntity implements TranslatableInterface compose TranslatebleTrait.

and here when Shopware implement

```php
/** @var EntityInterface $entity */
$entity = hydrate($meta, $data);

if($entity instanceof TranslatableInterface){
	//do translate
}
if($entity instanceof VersionableInterface){
	// do versioning
}
/** implement 3rd party api */
if($entity instanceof Symfony\Component\DependencyInjection\ContainerAwareInterface){
       $entity->setContainer($container);
}

return $entity;
```

This is also has benefit if Shopware has a plan to read definition/metadata from PHP attributes because it has no unnecessary property or method. We can put attributes to trait and compose to entity.

another pattern that benefit from interface 
```php
/** @var EntityInterface $entity */
$entity = hydrate($meta, $data);

$dispatcher->dispatch(new EntityCreatedEvent($entity));

return $entity;
```
TranslateEntitySubscriber.php
```php
public function onEntityCreated(EntityCreatedEvent $event)
{
   /** @var EntityInterface $entity */
   $entity = $event->getEntity();
   if(($entity instanceof TranslatableInterface) === false){
	return;
   }
  //do translate
}
```

**Comment by keulinho** [comment]
Sorry we won't go further in this direction, as all the interfaces etc. create a lot of public API surface area that can break, which will lead to tighter coupling and higher costs during updates.

**Reply by JoshuaBehrens** [reply]
I also see benefits of the event approach. It is less likely to break. We still need services, that can be replaced and parameters, that can be replaced. So it will be a discussion based approach which actions one want to "just" affect or replace
